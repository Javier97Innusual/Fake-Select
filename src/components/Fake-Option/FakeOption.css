.select-option {
    padding: 1rem;
    border-bottom: 2px solid black;
    font-weight: bold;
    cursor: pointer;

    &:last-child {
        border-bottom: none;
    }

    &.select-hidden {
        display: none;
    }
}