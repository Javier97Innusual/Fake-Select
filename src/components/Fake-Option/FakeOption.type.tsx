export type FakeOptionType = {
    value: string;
    children: string;
    hidden?: boolean;
};