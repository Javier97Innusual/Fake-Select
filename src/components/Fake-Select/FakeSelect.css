.select-main {
    display: flex;
    flex-flow: column;
    width: 400px;
    position: relative;

    .select-current {
        cursor: pointer;
        padding: 1.5rem 1rem;
        border: 2px solid black;
        font-size: 1.225rem;
        font-weight: bold;
        text-align: left;
    }

    .select-container {
        display: flex;
        flex-flow: column;
        border: 2px solid black;
        border-top: none;
    }

    &::after {
        content: '▼';
        position: absolute;
        right: 1rem;
        top: 1.725rem;
    }

    &.is-active {
        &::after {
            content: '✖';
        }
    }
}