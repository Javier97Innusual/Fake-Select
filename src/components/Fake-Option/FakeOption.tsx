import { FakeOptionType } from "./FakeOption.type";
import './FakeOption.css';

export function FakeOption({ value, children, hidden }: FakeOptionType) {
    return (
        <div
            className="select-option"
            data-id={value}
        >
            {children}
        </div>
    )
}