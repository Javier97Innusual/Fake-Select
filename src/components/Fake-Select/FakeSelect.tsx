import { MouseEvent, useState } from 'react';
import classNames from 'classnames';
import { DEFAULT_VALUE } from './FakeSelect.constants';
import { FakeSelectType, OptionValueType } from './FakeSelect.type';
import './FakeSelect.css';

export function FakeSelect({ id, placeholder, onChange, disabled, children }: FakeSelectType) {
    const [isSelectActive, setSelectActive] = useState<boolean>(false);
    const [currentOption, setCurrentOption] = useState<OptionValueType>({
        value: '',
        name: placeholder || DEFAULT_VALUE
    });

    const handleOption = (event: MouseEvent) => {
        const target = event.target as HTMLElement;

        if (target.dataset.id === undefined || target.innerHTML === null || disabled) {
            return;
        }

        setCurrentOption({
            value: target.dataset.id,
            name: target.innerText
        });

        onChange(target.dataset.id);

        setSelectActive(!isSelectActive);
    }

    return (
        <div
            id={id}
            className={classNames('select-main', { 'is-active': !disabled && isSelectActive })}
        >
            <div
                className="select-current"
                onClick={() => setSelectActive(!isSelectActive)}
            >
                {currentOption.name}
            </div>
            {!disabled && isSelectActive && (
                <>
                    <div
                        className="select-container"
                        onClick={handleOption}
                    >
                        {children}
                    </div>
                    <div
                        className="select-background"
                        onClick={() => setSelectActive(!isSelectActive)}
                    ></div>
                </>
            )}
        </div>
    )
}