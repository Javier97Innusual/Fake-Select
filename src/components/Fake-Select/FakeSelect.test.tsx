import { render, fireEvent } from '@testing-library/react';
import { FakeSelect } from './FakeSelect';
import { FakeOption } from './../Fake-Option/FakeOption';
import { DEFAULT_VALUE } from './FakeSelect.constants';

describe('FakeSelect component', () => {
    const onChangeMock = jest.fn();

    it('should render correctly and match snapshot', () => {
        const { asFragment } = render(
            <FakeSelect id="SelectName" onChange={onChangeMock}>
                <FakeOption value="red">Rojo</FakeOption>
                <FakeOption value="blue">Azul</FakeOption>
                <FakeOption value="green">Verde</FakeOption>
            </FakeSelect>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should open options list when clicked', () => {
        const { getByText } = render(
            <FakeSelect id="SelectName" onChange={onChangeMock}>
                <FakeOption value="red">Rojo</FakeOption>
                <FakeOption value="blue">Azul</FakeOption>
                <FakeOption value="green">Verde</FakeOption>
            </FakeSelect>
        );

        fireEvent.click(getByText(DEFAULT_VALUE));

        expect(getByText('Rojo')).toBeInTheDocument();
        expect(getByText('Azul')).toBeInTheDocument();
        expect(getByText('Verde')).toBeInTheDocument();
    });

    it('should update current option and closes list when an option is clicked', () => {
        const { getByText, container } = render(
            <FakeSelect id="SelectName" onChange={onChangeMock}>
                <FakeOption value="red">Rojo</FakeOption>
                <FakeOption value="blue">Azul</FakeOption>
                <FakeOption value="green">Verde</FakeOption>
            </FakeSelect>
        );

        fireEvent.click(getByText(DEFAULT_VALUE));
        fireEvent.click(getByText('Rojo'));

        expect(onChangeMock).toHaveBeenCalledWith('red');
        expect(container.querySelector('.select-container')).not.toBeInTheDocument();
    });

    it('should not open options list when disabled', () => {
        const { getByText, container } = render(
            <FakeSelect id="SelectName" onChange={onChangeMock} disabled>
                <FakeOption value="red">Rojo</FakeOption>
                <FakeOption value="blue">Azul</FakeOption>
                <FakeOption value="green">Verde</FakeOption>
            </FakeSelect>
        );

        fireEvent.click(getByText(DEFAULT_VALUE));

        expect(container.querySelector('.select-container')).not.toBeInTheDocument();
    });
});