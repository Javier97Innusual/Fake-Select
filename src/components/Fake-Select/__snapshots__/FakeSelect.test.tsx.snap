// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FakeSelect component should render correctly and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="select-main"
    id="SelectName"
  >
    <div
      class="select-current"
    >
      Elije una opci√≥n
    </div>
  </div>
</DocumentFragment>
`;
